### PR TITLE
allow treasury to do reserve asset transfers

### PR DESCRIPTION
**Pull request by samelamin**
This pr resolves https://github.com/paritytech/polkadot-sdk/issues/1428. 

*Added only to Kusama for now*

I did raise it [here](https://github.com/polkadot-fellows/runtimes/pull/19) and we discussed creating a chopsticks test to run an end-to-end test

however, to do that I will need a build agent/custom runner that is powerful enough to run the build

I will be doing that separately as I still think having chopsticks test your runtime with each commit will be very powerful and extremely useful for the ecosystem 

For now I have used XCM simulator and replicated what the other reserve tests do

### CODE REVIEW

**Comment by bkchr** [comment]
I don't get why you opened this here. The runtimes will be released from the fellowship repo. You also wanted to create a chopsticks test? 

TLDR: Please open against the fellowship repo.

**Comment by samelamin** [comment]
> I don't get why you opened this here. The runtimes will be released from the fellowship repo. You also wanted to create a chopsticks test?
> 
> TLDR: Please open against the fellowship repo.

Sorry, I was under the impression the fellowship repo is read-only till it gets migrated over. My mistake. 

I'll make the changes in that repo and close this pr

**Comment by samelamin** [comment]
reopening after discussion with Basti because this pr requires xcm changes that are not part of the fellowship repo currently

File: polkadot/xcm/xcm-builder/src/location_conversion.rs [hunk 1]
```diff
@@ -311,6 +311,8 @@ impl<Network: Get<Option<NetworkId>>, AccountId: From<[u8; 32]> + Into<[u8; 32]>
 			MultiLocation { parents: 0, interior: X1(AccountId32 { id, network }) }
 				if network == Network::get() =>
 				id,
+			MultiLocation { parents: 0, interior: X1(Plurality { id: BodyId::Treasury, .. }) } =>
+				PalletId(*b"py/trsry").into_account_truncating(),
```
Comment (franciscoaguirre) [inline polkadot/xcm/xcm-builder/src/location_conversion.rs:315]: You should probably add this on a new struct implementing `ConvertLocation`, since this one is for `AccountId32`s

**Comment by franciscoaguirre** [comment]
If I understand correctly, you'll leave the change to xcm-builder here, and change the runtimes in the runtimes repo?
And with that, you say you have to remove the test from here?
You could create a test for this with a mock runtime, no need to use the actual repos.

File: polkadot/xcm/xcm-builder/src/location_conversion.rs (hunk 1 above)
Comment (gavofyork) [reply polkadot/xcm/xcm-builder/src/location_conversion.rs:315]: ^^ That, except remove "probably".

File: polkadot/runtime/kusama/src/xcm_config.rs [hunk 2]
```diff
@@ -374,6 +378,9 @@ pub type StakingAdminToPlurality =
 /// Type to convert the Fellows origin to a Plurality `MultiLocation` value.
 pub type FellowsToPlurality = OriginToPluralityVoice<RuntimeOrigin, Fellows, FellowsBodyId>;
 
+/// Type to convert the Fellows origin to a Plurality `MultiLocation` value.
```
Comment (gavofyork) [inline polkadot/runtime/kusama/src/xcm_config.rs:381]:
```suggestion
/// Type to convert the Treasurer origin to a Plurality `MultiLocation` value.
```

**Comment by gavofyork** [comment]
Looks like you're mainly adding a means of converting Treasurer Frame-`Origin` to a local Plurality XCM-`Origin` and a means of converting a local Plurality XCM-`Origin` to a Frame-`Origin` of the Signed variant of the Treasury account ID.

Presumably you'll be wanting to communicate this XCM between two non-local XCM endpoints, in which case the local Plurality on the source endpoint will be a remote Plurality on the remote endpoint. In this case, your conversion to Treasury account ID won't work.

File: polkadot/xcm/xcm-builder/src/location_conversion.rs (hunk 1 above)
Comment (samelamin) [reply polkadot/xcm/xcm-builder/src/location_conversion.rs:315]: got it.

**Comment by samelamin** [comment]
> Looks like you're mainly adding a means of converting Treasurer Frame-`Origin` to a local Plurality XCM-`Origin` and a means of converting a local Plurality XCM-`Origin` to a Frame-`Origin` of the Signed variant of the Treasury account ID.
> 
> Presumably you'll be wanting to communicate this XCM between two non-local XCM endpoints, in which case the local Plurality on the source endpoint will be a remote Plurality on the remote endpoint. In this case, your conversion to Treasury account ID won't work.

Hey Gav, if I understand your concern, you are saying that this will only work from relay to para. 

We can resolve that by adding another match on `parent:1`. i.e.
```
		let id: [u8; 32] = match *location {
			MultiLocation { parents: 0, interior: X1(Plurality { id: BodyId::Treasury, .. }) } |
			MultiLocation { parents: 1, interior: X1(Plurality { id: BodyId::Treasury, .. }) } =>
				PalletId(*b"py/trsry").into_account_truncating(),
			_ => return None,
		};
```
Would that be acceptable?

**Comment by samelamin** [comment]
> If I understand correctly, you'll leave the change to xcm-builder here, and change the runtimes in the runtimes repo? And with that, you say you have to remove the test from here? You could create a test for this with a mock runtime, no need to use the actual repos.

Hey Cisco, thanks for the tip. Is there an existing test that uses MockRuntime that I should use to keep things consistent

Also I am assuming that if we want to isolate a test for only the xcm changes then all we need to do is test the location conversion right? because the "reserve_asset_transfer" is actually part of the runtime and can be included in a test there?

File: polkadot/xcm/xcm-builder/src/location_conversion.rs [hunk 3]
```diff
@@ -317,6 +317,22 @@ impl<Network: Get<Option<NetworkId>>, AccountId: From<[u8; 32]> + Into<[u8; 32]>
...
+/// Extracts the `AccountId32` from the passed Treasury plurality if the network matches.
+pub struct TreasuryVoiceConvertsVia<Network, AccountId>(PhantomData<(Network, AccountId)>);
+impl<Network: Get<Option<NetworkId>>, AccountId: From<[u8; 32]> + Into<[u8; 32]> + Clone>
+	ConvertLocation<AccountId> for TreasuryVoiceConvertsVia<Network, AccountId>
+{
+	fn convert_location(location: &MultiLocation) -> Option<AccountId> {
+		let id: [u8; 32] = match *location {
+			MultiLocation { parents: 0, interior: X1(Plurality { id: BodyId::Treasury, .. }) } |
+			MultiLocation { parents: 1, interior: X1(Plurality { id: BodyId::Treasury, .. }) } =>
+				PalletId(*b"py/trsry").into_account_truncating(),
```
Comment (bkchr) [inline polkadot/xcm/xcm-builder/src/location_conversion.rs:329]: IMO this should be generic. `TreasuryAccount: Get<AccountId>`

Comment (samelamin) [reply polkadot/xcm/xcm-builder/src/location_conversion.rs:329]: Sure I can do that, shall I keep the function in this file or move it somewhere else?

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-additional-tests
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3613847

File: polkadot/xcm/xcm-builder/src/location_conversion.rs (hunk 3 above)
Comment (bkchr) [reply polkadot/xcm/xcm-builder/src/location_conversion.rs]: Which function?

Comment (samelamin) [reply polkadot/xcm/xcm-builder/src/location_conversion.rs]: disregard, i used TreasuryAccount and passing it in from the runtime/xcm_config

File: polkadot/xcm/xcm-builder/src/location_conversion.rs [hunk 4]
```diff
@@ -317,6 +317,27 @@ impl<Network: Get<Option<NetworkId>>, AccountId: From<[u8; 32]> + Into<[u8; 32]>
...
+		Network: Get<Option<NetworkId>>,
+		AccountId: From<[u8; 32]> + Into<[u8; 32]> + Clone,
+		TreasuryAccount: Get<AccountId>,
+	> ConvertLocation<AccountId> for TreasuryVoiceConvertsVia<Network, AccountId, TreasuryAccount>
+{
+	fn convert_location(location: &MultiLocation) -> Option<AccountId> {
+		let id: [u8; 32] = match *location {
+			MultiLocation { parents: 0, interior: X1(Plurality { id: BodyId::Treasury, .. }) } |
+			MultiLocation { parents: 1, interior: X1(Plurality { id: BodyId::Treasury, .. }) } =>
+				TreasuryAccount::get().into(),
```
Comment (franciscoaguirre) [inline polkadot/xcm/xcm-builder/src/location_conversion.rs:334]: This should be done via `HashedDescription` for `parents: 1`, can be kept local for `parents: 0`

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-additional-tests
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3663848

File: polkadot/xcm/xcm-builder/src/location_conversion.rs [hunk 5]
```diff
@@ -317,6 +317,46 @@ impl<Network: Get<Option<NetworkId>>, AccountId: From<[u8; 32]> + Into<[u8; 32]>
...
+/// Create a description of the remote treasury `location` if possible. No two locations should have
+/// the same descriptor.
+pub struct DescribeTreasuryVoice;
+
+impl DescribeLocation for DescribeTreasuryVoice {
+	fn describe_location(location: &MultiLocation) -> Option<Vec<u8>> {
+		Some(match location {
+			// Used on the relay chain for sending paras that use 32 byte accounts
+			MultiLocation {
+				parents: 1,
```
Comment (franciscoaguirre) [inline polkadot/xcm/xcm-builder/src/location_conversion.rs:329]: You could leave this as a terminal treasury voice, `parents: 0`, and reuse `DescribeFamily`

File: polkadot/xcm/xcm-builder/src/location_conversion.rs [hunk 6]
```diff
@@ -435,6 +475,9 @@ mod tests {
 	pub type ForeignChainAliasAccount<AccountId> =
 		HashedDescription<AccountId, LegacyDescribeForeignChainAccount>;
 
+	pub type ForeignChainAliasTreasuryAccount<AccountId> =
+		HashedDescription<AccountId, DescribeTreasuryVoice>;
```
Comment (franciscoaguirre) [inline polkadot/xcm/xcm-builder/src/location_conversion.rs:479]: Do you need this here? You already have it on the parent crate

**Review by franciscoaguirre — APPROVED** [review]

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3701431

**Comment by samelamin** [comment]
@bkchr @KiChjang is there anything else you want changed here (apart from conflicts!)

**Review by KiChjang — APPROVED** [review]

**Comment by samelamin** [comment]
With the merging of this [pr](https://github.com/paritytech/polkadot-sdk/pull/1731) i had to remove the changes made to the runtime to allow the treasury to send xcm

Once this is merged, i shall make a separate pr to the fellowship repo with the xcm changes to kusama

File: cumulus/parachains/integration-tests/emulated/common/src/constants.rs [hunk 7]
```diff
@@ -70,6 +70,10 @@ pub mod accounts {
 	pub const FERDIE_STASH: &str = "Ferdie//stash";
 	pub const FERDIE_BEEFY: &str = "Ferdie//stash";
 
+	pub fn get_treasury_address() -> AccountId {
+		PalletId(*b"py/trsry").into_account_truncating()
```
Comment (kianenigma) [inline cumulus/parachains/integration-tests/emulated/common/src/constants.rs:74]: is there no possible way to _NOT_ hardcode this?

File: polkadot/xcm/xcm-builder/src/location_conversion.rs [hunk 8]
```diff
@@ -84,6 +84,20 @@ impl DescribeLocation for DescribeAccountKey20Terminal {
 	}
 }
 
+/// Create a description of the remote treasury `location` if possible. No two locations should have
+/// the same descriptor.
+pub struct DescribeTreasuryVoiceTerminal;
```
Comment (kianenigma) [inline polkadot/xcm/xcm-builder/src/location_conversion.rs:89]: very cute name.

**Review by kianenigma — COMMENTED** [review]
would be good to be able to see an end to end example of this that is actually about the treasury sending a reserved asset transfer.

File: cumulus/parachains/integration-tests/emulated/common/src/constants.rs (hunk 7 above)
Comment (samelamin) [reply cumulus/parachains/integration-tests/emulated/common/src/constants.rs:74]: Its hardcoded in the runtime too, we can make it a constant though?

**Comment by samelamin** [comment]
> would be good to be able to see an end to end example of this that is actually about the treasury sending a reserved asset transfer.

We [did](https://github.com/paritytech/polkadot-sdk/pull/1447/commits/f97b8b48c6e03f3134752f3165b6805deee672d6#diff-9b70f65c316d861a1651852aad6b780bc0dfa5f1ea585f008c91a5135bb4d4bfR448) have an integration test but as I mentioned [previously](https://github.com/paritytech/polkadot-sdk/pull/1447#issuecomment-1740525041), the removal of the runtimes to the fellowship means we will need to do the test there after this pr is merged

File: cumulus/parachains/integration-tests/emulated/common/src/constants.rs (hunk 7 above)
Comment (kianenigma) [reply cumulus/parachains/integration-tests/emulated/common/src/constants.rs:74]: Yes please.

**Comment by samelamin** [comment]
@kianenigma @bkontur @NachoPal @bkchr any more changes required on this pr?

File: polkadot/xcm/xcm-builder/src/location_conversion.rs [hunk 9]
```diff
@@ -936,4 +973,22 @@ mod tests {
 		};
 		assert!(ForeignChainAliasAccount::<[u8; 32]>::convert_location(&mul).is_none());
 	}
+
+	#[test]
+	fn remote_account_convert_on_para_sending_relay_treasury() {
+		let mul = MultiLocation {
```
Comment (bkontur) [inline polkadot/xcm/xcm-builder/src/location_conversion.rs:979]: just a very nit, I would rename `let mul` to `let loc` or `let location`. E.g. [xcm-v4](https://github.com/paritytech/polkadot-sdk/pull/1230) renames `MultiLocation` to `Location`, so `let mul = Location...` will look strange

Comment (samelamin) [reply polkadot/xcm/xcm-builder/src/location_conversion.rs]: resolved

File: polkadot/xcm/xcm-builder/src/location_conversion.rs [hunk 10]
```diff
@@ -328,6 +343,25 @@ impl<Network: Get<Option<NetworkId>>, AccountId: From<[u8; 32]> + Into<[u8; 32]>
 	}
 }
 
+/// Extracts the `AccountId32` from the passed Treasury plurality if the network matches.
```
Comment (bkontur) [inline polkadot/xcm/xcm-builder/src/location_conversion.rs:346]:
```suggestion
/// Returns specified `TreasuryAccount` as `AccountId32` if passed `location` matches Treasury plurality.
```

I dont see any network matching here, I assume this is kind of copy/paste from `AccountId32Aliases` :)

File: polkadot/xcm/xcm-builder/src/location_conversion.rs [hunk 11]
```diff
@@ -328,6 +343,25 @@ impl<Network: Get<Option<NetworkId>>, AccountId: From<[u8; 32]> + Into<[u8; 32]>
 	}
 }
 
+/// Returns specified `TreasuryAccount` as `AccountId32` if passed `location` matches Treasury plurality.
+pub struct LocalTreasuryVoiceConvertsVia<TreasuryAccount, AccountId>(
```
Comment (bkontur) [inline polkadot/xcm/xcm-builder/src/location_conversion.rs:347]: it would be nice to add here simple test for this converter

File: polkadot/xcm/xcm-builder/src/location_conversion.rs [hunk 12]
```diff
@@ -328,6 +343,25 @@ impl<Network: Get<Option<NetworkId>>, AccountId: From<[u8; 32]> + Into<[u8; 32]>
...
+	PhantomData<(TreasuryAccount, AccountId)>,
+);
+impl<TreasuryAccount: Get<AccountId>, AccountId: From<[u8; 32]> + Into<[u8; 32]> + Clone>
+	ConvertLocation<AccountId> for LocalTreasuryVoiceConvertsVia<TreasuryAccount, AccountId>
+{
+	fn convert_location(location: &MultiLocation) -> Option<AccountId> {
+		let id: [u8; 32] = match *location {
```
Comment (bkontur) [inline polkadot/xcm/xcm-builder/src/location_conversion.rs:354]:
```suggestion
        match *location {
			MultiLocation {
				parents: 0,
				interior: X1(Plurality { id: BodyId::Treasury, part: BodyPart::Voice }),
			} => Some(TreasuryAccount::get().into()),
			_ => None,
		}
```

I think it should  be possible without temp variable

File: polkadot/xcm/xcm-builder/src/location_conversion.rs [hunk 13]
```diff
@@ -84,6 +84,20 @@ impl DescribeLocation for DescribeAccountKey20Terminal {
...
 
+/// Create a description of the remote treasury `location` if possible. No two locations should have
+/// the same descriptor.
+pub struct DescribeTreasuryVoiceTerminal;
+
+impl DescribeLocation for DescribeTreasuryVoiceTerminal {
+	fn describe_location(l: &MultiLocation) -> Option<Vec<u8>> {
+		match (l.parents, &l.interior) {
+			(0, X1(Plurality { id: BodyId::Treasury, part: BodyPart::Voice })) =>
+				Some(("Treasury").encode()),
```
Comment (bkontur) [inline polkadot/xcm/xcm-builder/src/location_conversion.rs:95]: I am not sure, is this enough?
If I check `DescribeBodyTerminal`, it use `Some((b"Body", id, part)`

```suggestion
				Some((b"Treasury", b"Voice").encode()),
```

File: cumulus/parachains/integration-tests/emulated/common/src/constants.rs (hunk 7 above)
Comment (bkontur) [reply cumulus/parachains/integration-tests/emulated/common/src/constants.rs:74]: I think we dont need to add this here,
it is only used for `init_balances` where we drip some common accounts.

I think if any test needs treasury account with some balance, it should be as a part of preparing data for that concrete test. And then that concrete test can use runtime-specific Treasury account.

So I guess we dont need to specify `PalletId(*b"py/trsry")` here and also we dont need to even touch integration-tests in this PR. So I would revert changes from this file.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3915033

File: polkadot/xcm/xcm-builder/src/location_conversion.rs [hunk 14]
```diff
@@ -328,6 +343,31 @@ impl<Network: Get<Option<NetworkId>>, AccountId: From<[u8; 32]> + Into<[u8; 32]>
...
+{
+	fn convert_location(location: &MultiLocation) -> Option<AccountId> {
+        match *location {
+			MultiLocation {
+				parents: 0,
+				interior: X1(Plurality { id: BodyId::Treasury, part: BodyPart::Voice }),
+			} => Some(TreasuryAccount::get().into()),
+			_ => None,
+		}
+			MultiLocation {
```
Comment (bkontur) [inline polkadot/xcm/xcm-builder/src/location_conversion.rs:361]: @samelamin sorry, for my bad suggestion, please remove these bellow

File: polkadot/xcm/xcm-builder/src/location_conversion.rs [hunk 15]
```diff
@@ -84,6 +84,20 @@ impl DescribeLocation for DescribeAccountKey20Terminal {
...
+/// Create a description of the remote treasury `location` if possible. No two locations should have
+/// the same descriptor.
+pub struct DescribeTreasuryVoiceTerminal;
+
+impl DescribeLocation for DescribeTreasuryVoiceTerminal {
+	fn describe_location(l: &MultiLocation) -> Option<Vec<u8>> {
+		match (l.parents, &l.interior) {
+			(0, X1(Plurality { id: BodyId::Treasury, part: BodyPart::Voice })) =>
+				Some((b"Treasury", b"Voice").encode()),
+			_ => return None,
```
Comment (bkontur) [inline polkadot/xcm/xcm-builder/src/location_conversion.rs:96]:
```suggestion
			_ => None,
```

File: polkadot/xcm/xcm-builder/src/location_conversion.rs [hunk 16]
```diff
@@ -936,4 +979,22 @@ mod tests {
 		};
 		assert!(ForeignChainAliasAccount::<[u8; 32]>::convert_location(&mul).is_none());
 	}
+
+	#[test]
+	fn remote_account_convert_on_para_sending_relay_treasury() {
```
Comment (bkontur) [inline polkadot/xcm/xcm-builder/src/location_conversion.rs:984]: probably I am missing some context here, I can understand `remote_account_convert_on_para` - that this converter will be used as a part of `LocationToAccountId` converter on parachain,
but what does exactly mean `sending_relay_treasury`?

Comment (samelamin) [reply polkadot/xcm/xcm-builder/src/location_conversion.rs:984]: its a badly named test i think, but its meant to be remote treasury on a parachain to another parachain

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3915688

File: cumulus/parachains/integration-tests/emulated/common/src/constants.rs (hunk 7 above)
Comment (samelamin) [reply cumulus/parachains/integration-tests/emulated/common/src/constants.rs]: removed in latest commit

File: polkadot/xcm/xcm-builder/src/location_conversion.rs (hunk 16 above)
Comment (bkontur) [reply polkadot/xcm/xcm-builder/src/location_conversion.rs:978]: hmm, because from the **parachain's point of view** this:
``` 
        MultiLocation {
			parents: 1,
			interior: X1(Plurality { id: BodyId::Treasury, part: BodyPart::Voice }),
		};
```
represents Plurality-Treasury-Voice location on relay chain (because `parents: 1`).


From the **parachain's point-of-view** a **relay chain plurality** looks like:
```
MultiLocation { parents: 1, interior: X1(Plurality { .. }
```

From the **parachain's point-of-view** an **other parachain plurality** looks like:
```
MultiLocation { parents: 1, interior: X2(Parachain(other_parachain_para_id), Plurality { .. }
```

So I would suggest to cover at least these two case in this test, to be sure that it works as expected.
Ideally, it should maybe cover more cases:
DescribeTreasuryVoiceTerminal is deployed on relay chain and local treasury location.
DescribeTreasuryVoiceTerminal is deployed on relay chain and  parachain treasury location? (not sure about this case)
DescribeTreasuryVoiceTerminal is deployed on parachain and local treasury location.
DescribeTreasuryVoiceTerminal is deployed on parachain and relay chain treasury location.
DescribeTreasuryVoiceTerminal is deployed on parachain and other parachain treasury location.

(but as I said, I dont have full context here)

Comment (samelamin) [reply polkadot/xcm/xcm-builder/src/location_conversion.rs]: Hi @bkontur thanks for this, I've renamed the tests and added another test to cover the para to para treasury conversion



I _think_ the `DescribeFamily` covers all the other cases but please correct me if I am mistaken.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3919982

File: polkadot/xcm/xcm-builder/src/lib.rs [hunk 17]
```diff
@@ -34,8 +34,8 @@ pub use location_conversion::{
 	ChildParachainConvertsVia, DescribeAccountId32Terminal, DescribeAccountIdTerminal,
 	DescribeAccountKey20Terminal, DescribeAllTerminal, DescribeBodyTerminal, DescribeFamily,
```
Comment (bkontur) [inline polkadot/xcm/xcm-builder/src/lib.rs:35]: I guess, you are missing re-export for `DescribeTreasuryVoiceTerminal` here.y You wont be able to use it outside of xcm-builder without that.

**Review by bkontur — APPROVED** [review]
please, revert also those imports from `cumulus/parachains/integration-tests/emulated/common/src/constants.rs`

**Comment by samelamin** [comment]
> please, revert also those imports from `cumulus/parachains/integration-tests/emulated/common/src/constants.rs`

reverted
